src/Workspaces/Remote/ServiceHub/Host/RemoteSolutionCache.cs

```diff
@@ -0,0 +1,161 @@
+    /// collect useful telemetry on how much benefit we would get by having a larger history.
+    /// </summary>
+    private readonly LinkedList<CacheNode> _cacheNodes = new();
+
+    public RemoteSolutionCache(int maxCapacity = 4, int totalHistory = 16)
+    {
+        _maxCapacity = maxCapacity;
+        _totalHistory = totalHistory;
+        _cacheMissCounter = new(bucketSize: 1, maxBucketValue: int.MaxValue, bucketCount: _totalHistory + 1);
+    }
+
+    private void FindAndMoveNodeToFront(TChecksum checksum)
+    {
+        var index = 0;
+        for (var current = _cacheNodes.First; current != null; current = current.Next, index++)
+        {
+            if (current.Value.Checksum.Equals(checksum))
+            {
+                // Found the item.  Take it, move it to the front.
+                _cacheNodes.Remove(current);
+                _cacheNodes.AddFirst(current);
+                return;
+            }
+        }
+
+        // Didn't find the item at all.  Just add to the front.
+        _cacheNodes.AddFirst(new CacheNode(checksum));
+    }
+
+    public void Add(TChecksum checksum, TSolution solution)
+    {
+        Contract.ThrowIfTrue(_cacheNodes.Count > _totalHistory);
+
+        FindAndMoveNodeToFront(checksum);
+
+        Contract.ThrowIfTrue(_cacheNodes.Count > _totalHistory + 1);
+        Contract.ThrowIfNull(_cacheNodes.First);
+        Contract.ThrowIfFalse(_cacheNodes.First.Value.Checksum.Equals(checksum));
+
+        // Ensure we're holding onto the solution.
+        _cacheNodes.First.Value.Solution = solution;
+
+        // Now, if our history is too long, remove the last item.
+        if (_cacheNodes.Count == _totalHistory + 1)
+            _cacheNodes.RemoveLast();
+
+        // Finally, ensure that only the first `MaxCapacity` are pointing at solutions and the rest are not.
+        var index = 0;
+        for (var current = _cacheNodes.First; current != null; current = current.Next, index++)
+        {
+            if (index >= _maxCapacity)
+            {
+                current.Value.Solution = null;
+                break;
+            }
+        }
+    }
+
+    public TSolution? Find(TChecksum checksum)
+    {
+        // Note: we intentionally do not move an item when we find it.  That's because our caller will always call 'Add'
+        // with the found solution afterwards.  This will ensure that that solution makes it to the front of the line.
+        var index = 0;
+        for (var current = _cacheNodes.First; current != null; current = current.Next, index++)
+        {
+            if (current.Value.Checksum.Equals(checksum))
+            {
+                // Found it!
+                if (current.Value.Solution is null)
+                {
+                    // Track that we would have been able to return this if our history was longer.
+                    _cacheMissesInHistory++;
+                    _cacheMissCounter.IncreaseCount(index);
```

Comment:
>_cacheMissCounter.IncreaseCount(index); [](http://example.com/codeflow?start=20&length=39)

I guess it's confusing to me why this isn't done in the success case too (I'd also have thought the name wouldn't indicate a miss)

Reply:
OK, I guess codeflow had a caching problem, because I see the change now. Never mind the above comment.